app/scripts/modules/amazon/src/serverGroup/configure/wizard/pages/fields/ServerGroupDetailsField.tsx

```diff
@@ -0,0 +1,229 @@
+ * spaces, dots etc. These incorrect values will get combined to form the consolidated value for `freeformDetails`
+ * field. This consolidated value will be parsed again when `ServerGroupDetailsField` is re-rendered, but the parsing
+ * will fail this time since the consolidated value does not conform to the expected format. This will result in a
+ * state where the error cannot be shown to the user and won't be provided a way to fix it.
+ *
+ * Keeping a local error state allows us to bail out from informing the parent about the change and provide the user
+ * a way to fix the error.
+ */
+function useLocalError(
+  validator: (value: string) => string | null,
+  successHandler: (value: string) => void,
+  errorHandler: (error: string) => void,
+): [LocalError | null, (value: string) => void] {
+  const [localError, setLocalError] = useState<LocalError>(null);
+  const changeHandler = useCallback(
+    (value: string) => {
+      const error = validator(value);
+      if (!error) {
+        successHandler(value);
+        setLocalError(null);
+      } else {
+        setLocalError({ error, value });
+        errorHandler(error);
+      }
+    },
+    [validator, successHandler],
+  );
+
+  return [localError, changeHandler];
+}
+
+interface ShardFieldProps {
+  shardIndex: number;
+  value: string;
+  onChange: (shardIndex: number, value: string) => void;
+  onError: (error: string) => void;
+}
+function ShardField({ shardIndex, value, onChange, onError }: ShardFieldProps) {
+  const [localError, changeHandler] = useLocalError(
+    validateShard,
+    (value: string) => {
+      onChange(shardIndex, value);
+    },
+    onError,
+  );
+
+  const onValueChange = useCallback(
+    (e: SyntheticEvent) => {
+      changeHandler((e.target as HTMLInputElement).value);
+    },
+    [changeHandler],
+  );
+
+  return (
+    <>
+      <div className="form-group">
+        <div className="col-md-3 sm-label-right">
+          Shard {shardIndex} <HelpField id="aws.serverGroup.shard" />
+        </div>
+        <div className="col-md-7">
+          <input
+            type="text"
+            className="form-control input-sm no-spel"
+            // When there is an error, the parent wasn't informed of the change, so use the value from local error
+            // state.
+            value={localError ? localError.value : value}
+            onChange={onValueChange}
+          />
+        </div>
+      </div>
+      {localError && (
+        <div className="form-group row slide-in">
+          <div className="col-md-7 col-md-offset-3 error-message">
+            <span>{localError.error}</span>
+          </div>
+          <span></span>
+        </div>
+      )}
+    </>
+  );
+}
+
+export interface OtherDetailsFieldProps {
+  value: string;
+  onChange: (value: string) => void;
+  onError: (error: string) => void;
+}
+function OtherDetailsField({ value, onChange, onError }: OtherDetailsFieldProps) {
```

Comment:
The `ShardField` and the `OtherDetailsField` have the same behavior and similar return values. We could make this more reusable by extracting both components to a generic `InputWithLocalError` that accepts the validator function, label, and help text as props (in addition to the existing props).